### PR TITLE
Document Service Traffic Policies

### PR DESCRIPTION
**Pull request by andrewsykim**
Signed-off-by: Andrew Sy Kim kim.andrewsy@gmail.com

Documents traffic policies that can be set for Services. This PR adds documentation for both internalTrafficPolicy and externalTrafficPolicy. externalTrafficPolicy has always existed and is not new, but I realized the services document didn't have any explicit mention of it so I added it here. internalTrafficPolicy is a new feature we promoted to beta in v1.22. This PR also documents the new ProxyTerminatingEndpoints feature (alpha) and it's relation to externalTrafficPolicy.

Related KEPS:
kubernetes/enhancements#2086
kubernetes/enhancements#1669

Original PR was https://github.com/kubernetes/website/pull/28981

### CODE REVIEW

File: content/en/docs/concepts/services-networking/service.md [hunk 1]
```diff
@@ -384,6 +384,33 @@ The IP address that you choose must be a valid IPv4 or IPv6 address from within
 If you try to create a Service with an invalid clusterIP address value, the API
 server will return a 422 HTTP status code to indicate that there's a problem.
 
+## Traffic Policies
+
+### External Traffic Policy
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/service.md:389]: style nit: this should be sentence case
```suggestion
### External traffic policy
```

File: content/en/docs/concepts/services-networking/service.md [hunk 2]
```diff
@@ -384,6 +384,33 @@ The IP address that you choose must be a valid IPv4 or IPv6 address from within
...
+{{< note >}}
+{{< feature-state for_k8s_version="v1.22" state="alpha" >}}
+If the feature gate ProxyTerminatingEndpoints is enabled from kube-proxy, it will route traffic to terminating endpoints
+if and only if the external traffic policy is `Local` and there are only terminating endpoints on the node. This capability
+was added to allow external load balancers to gracefully drain connections from node ports even when the health check node
+port starts to fail. Otherwise, traffic can be lost between the time a node is still in the node pool of a load balancer
+and traffic is being dropped during the termination period of a pod.
+{{< /note >}}
+
+### Internal Traffic Policy
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/service.md:405]: style nit: this should be sentence case
```suggestion
### Internal traffic policy
```

File: content/en/docs/concepts/services-networking/service.md [hunk 3]
```diff
@@ -384,6 +384,33 @@ The IP address that you choose must be a valid IPv4 or IPv6 address from within
 If you try to create a Service with an invalid clusterIP address value, the API
 server will return a 422 HTTP status code to indicate that there's a problem.
 
+## Traffic Policies
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/service.md:387]: style nit: this should be sentence case
```suggestion
## Traffic policies
```

File: content/en/docs/concepts/services-networking/service.md [hunk 4]
```diff
@@ -384,6 +384,33 @@ The IP address that you choose must be a valid IPv4 or IPv6 address from within
...
+
+### External Traffic Policy
+
+You can set the `spec.externalTrafficPolicy` field to control how traffic from external sources is routed.
+Valid values are `Cluster` and `Local`. Set the field to `Cluster` to route external traffic to all ready endpoints
+and `Local` to only route to ready node-local endpoints. If the traffic policy is `Local` and there are are no node-local
+endpoints, traffic is dropped by kube-proxy.
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/service.md:394]:
```suggestion
endpoints, the kube-proxy does not forward any traffic for the relevant Service.
```
I presume the exact behavior in this case depends on whether you're using kube-proxy in ipvs, iptables or userland mode. Is that right?

File: content/en/docs/concepts/services-networking/service.md [hunk 5]
```diff
@@ -384,6 +384,33 @@ The IP address that you choose must be a valid IPv4 or IPv6 address from within
...
+
+{{< note >}}
+{{< feature-state for_k8s_version="v1.22" state="alpha" >}}
+If the feature gate ProxyTerminatingEndpoints is enabled from kube-proxy, it will route traffic to terminating endpoints
+if and only if the external traffic policy is `Local` and there are only terminating endpoints on the node. This capability
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/service.md:399]:
```suggestion
If you enable the `ProxyTerminatingEndpoints`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
`ProxyTerminatingEndpoints` for the kube-proxy, the kube-proxy checks if the node
has local endpoints and whether or not all the local endpoints are marked as terminating.

If there are local endpoints and **all** of those are terminating, then the kube-proxy ignores
any external traffic policy of `Local`. Instead, whilst the node-local endpoints remain as all
terminating, the kube-proxy forwards traffic for that Service to healthy endpoints elsewhere,
as if the external traffic policy were set to `Cluster`.

This forwarding behavior for terminating endpoints
```
```

File: content/en/docs/concepts/services-networking/service.md [hunk 6]
```diff
@@ -384,6 +384,33 @@ The IP address that you choose must be a valid IPv4 or IPv6 address from within
...
+was added to allow external load balancers to gracefully drain connections from node ports even when the health check node
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/service.md:400]:
```suggestion
exists to allow external load balancers to gracefully drain connections that are backed by
`NodePort` Services, even when the health check node
```

**Review by sftim — COMMENTED** [review]
Style LGTM. I've suggested some wording tweaks.

**Comment by PI-Victor** [comment]
@andrewsykim can you please address the comments above so we can move forward, the deadline for PRs has now passed, thanks!

**Comment by sftim** [comment]
To be honest, this is OK to merge even with the nits left in, _provided_ that it passes tech review from SIG Network.

**Comment by sftim** [comment]
@kubernetes/sig-network-pr-reviews could you check this one out?

**Comment by PI-Victor** [comment]
ping @kubernetes/sig-network-pr-reviews again for review
@andrewsykim can you please address the comments above so we can merge this today, thanks!

**Comment by PI-Victor** [comment]
truly **infuriating**
/close
in favor of https://github.com/kubernetes/website/pull/29156

**Pull request closed without merging**